### PR TITLE
Add ERC20Channels contract with test and docs

### PR DESCRIPTION
**Pull request by AugustoL**
The ERC20Channels contract is a simple implementation of state channels for ERC20 tokens. This contract allows the token holders to mediate and decide about how much the final value of a transaction is going to be.
The channel is opened by a token holder that previously approved the transfer of tokens to the ERC20Channels contract, once the channel is opened the tokens are transfered to the contract and the mediation of the funds start.
While the channel is opened the two parties can exchange signatures proving the allowance and claim of the funds to be transfered, in this case the sender will share signatures where he allows the receiver to claim certain amount of tokens and the receiver will share signatures where he is willing to claim that amount.
With those signatures the receiver and sender can reach a mutual agreement on how much the final value of the transfer is going to be and always the time to prevent the another party to close the channel with the wrong final value.

Based in [LifChannels contract](https://github.com/windingtree/LifToken/blob/master/contracts/LifChannels.sol)

### CODE REVIEW

File: contracts/token/ERC20Channels.sol [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+  function cooperativeClose(
+    address receiver,
+    uint8 nonce,
+    uint256 balance,
+    bytes balanceMsgSig,
+    bytes closingSig
+  ) external {
+    // Derive receiver address from signature
+    bytes32 msgHash = keccak256(balanceMsgSig);
+    require(receiver == msgHash.recover(closingSig));
```
[the remark this answers is not in the document]
Comment (spalladino) [reply contracts/token/ERC20Channels.sol]: Also, should we consider adding a variant of this method where the receiver is actually the `msg.sender`, so it does not require an offline signature of `balanceMsgSig`? If I understand correctly, this will typically be called by the `receiver`, so we could simplify the flow for this case.

File: test/ERC20Channel.test.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  it('create channel and close it from sender with uncooperativeClose', async function () {
+    await token.approve(tokenChannels.address, 30, { from: accounts[2] });
+    const nonce = 33;
+    await tokenChannels.openChannel(accounts[1], 30, nonce, { from: accounts[2] });
+    const hash = await tokenChannels.generateBalanceHash(accounts[1], nonce, 10);
+    const senderSig = signMsg(hash, privateKeys[accounts[2]]);
+    assert.equal(accounts[2],
+      await tokenChannels.getSignerOfBalanceHash(accounts[1], nonce, 10, senderSig)
+    );
```
Comment (spalladino) [inline test/ERC20Channel.test.js]: Remove the `senderSig` from this test as it is not used (copy-paste?)

File: test/ERC20Channel.test.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it('create channel', async function () {
+    await token.approve(tokenChannels.address, 30, { from: accounts[2] });
+    const nonce = 66;
+    await tokenChannels.openChannel(accounts[1], 30, nonce, { from: accounts[2] });
+    const channelInfo = await tokenChannels.getChannelInfo(accounts[2], accounts[1], nonce);
+    assert.equal(channelInfo[1], 30);
+    assert.equal(channelInfo[2], 0);
+    assert.equal(channelInfo[3], 0);
```
Comment (spalladino) [inline test/ERC20Channel.test.js]: Could you add an assertion that the tokens are actually transferred from the sender to the channels contract? And maybe another test in which creation fails if the sender had not approved the transfer previously?

File: test/ERC20Channel.test.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    );
+    const senderHash = await tokenChannels.generateKeccak256(senderSig);
+    const closingSig = signMsg(senderHash, privateKeys[accounts[1]]);
+    await tokenChannels.uncooperativeClose(accounts[1], nonce, 10, { from: accounts[2] });
+    await increaseTimeTo(latestTime() + 10);
+    await tokenChannels.cooperativeClose(accounts[1], nonce, 20, senderSig, closingSig, { from: accounts[1] });
+    (await token.balanceOf(accounts[2])).should.be.bignumber
+      .equal(80);
+    (await token.balanceOf(accounts[1])).should.be.bignumber
+      .equal(20);
```
Comment (spalladino) [inline test/ERC20Channel.test.js]: Could you add an assertion, in all tests that end with a successful close, that the call to `getChannelInfo` returns empty?

File: contracts/token/ERC20Channels.sol [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    bytes32 channelId = getChannelId(sender, receiver, nonce);
+    Channel memory channel = channels[channelId];
+
+    require(channel.nonce > 0);
+    require(receiverBalance <= channel.deposit);
+
+    // Remove closed channel structures
+    // channel.nonce will become 0
+    // Change state before transfer call
+    delete channels[channelId];
```
Comment (spalladino) [inline contracts/token/ERC20Channels.sol]: Is it safe to reset `channel.nonce` to 0 after closing the contract? It would allow another channel to be created with the same nonce, for which the previous set of offline signed messages are still valid. Though only a `sender` can create a channel, and he should not do this as it is only in his own prejudice (as it allows the receiver to use the previous signed messages to extract the balance), should we add a safeguard to prevent the sender from that doing that mistake?

File: test/ERC20Channel.test.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  .use(require('chai-bignumber')(BigNumber))
+  .should();
+
+const assertRevert = require('./helpers/assertRevert');
+
+var StandardTokenMock = artifacts.require('./mocks/StandardTokenMock.sol');
+var ERC20Channels = artifacts.require('./token/ERC20Channels.sol');
+var ECRecovery = artifacts.require('./ECRecovery.sol');
+
+contract('ERC20Channels', function (accounts) {
```
Comment (spalladino) [inline test/ERC20Channel.test.js]: Could you add tests for trying to close a channel (both cooperatively and uncooperatively) with a higher balance than the original deposit, _with the contract actually having that balance at its disposal_ (due to other channels being open as well)?

**Comment by maraoz** [comment]
+1 to the "spawn a contract per channel" approach. It seems simpler and harder to break. I'd rename the contract to `ERC20Channel`, and maybe later we can add a `ERC20MultiChannel`.

Great contribution @AugustoL !!! :1st_place_medal:

File: contracts/token/ERC20Channels.sol [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    uint8 nonce;
+  }
+
+  // The requests from sender to close a channel
+  mapping (bytes32 => ClosingRequest) public closingRequests;
+  struct ClosingRequest {
+    uint256 closingBalance;
+    uint8 state;
+    uint256 closeTime;
+  }
```
[the remark this answers is not in the document]
Comment (AugustoL) [reply contracts/token/ERC20Channels.sol]: I think this way the contract storage will be smaller, it will store only the information necessary at any stage, if it is open only the channel information and once it want to be closed it will also store a closeRequest. @spalladino do you know if working with the structs like this will really optimize the contract storage ? Maybe im wrong!

File: test/ERC20Channel.test.js (hunk 6 above)
Comment (AugustoL) [reply test/ERC20Channel.test.js]: yep!

File: test/ERC20Channel.test.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+  beforeEach(async function () {
+    const ecrecovery = await ECRecovery.new();
+    ERC20Channels.link('ECRecovery', ecrecovery.address);
+    token = await StandardTokenMock.new(accounts[2], 100);
+    tokenChannels = await ERC20Channels.new(token.address, 60);
+  });
+
+  it('create channel', async function () {
+    await token.approve(tokenChannels.address, 30, { from: accounts[2] });
```
[the remark this answers is not in the document]
Comment (AugustoL) [reply test/ERC20Channel.test.js]: :+1:

File: test/ERC20Channel.test.js (hunk 3 above)
Comment (AugustoL) [reply test/ERC20Channel.test.js]: :+1:

**Comment by AugustoL** [comment]
Thx for the review @maraoz and @spalladino !

I agree that it will be better to have a ERC20Channel contract and use once contract per channel.

My to do list:

- Rename ERC20Channels to ERC20Channel and change the contract to work only as one channel.
- Add another method to close a channel cooperatively from receiver.
- Add more tests.

File: contracts/token/ERC20Channels.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (spalladino) [reply contracts/token/ERC20Channels.sol]: As far as I understand, storing a value of zero in storage has no gas cost, so it should make no difference. But I could be wrong, I think it's worth testing.

File: contracts/token/ERC20Channels.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (spalladino) [reply contracts/token/ERC20Channels.sol]: Could be. Or maybe an overload that does not require the extra signature? Either way, since you have more actual experience in using this contract, I'll leave it up to you whether it's worthwhile to implement such method or not.

File: test/ERC20Channel.test.js (hunk 6 above)
Comment (spalladino) [reply test/ERC20Channel.test.js]: Actually, if we split the channels so there is one contract per channel, this test would no longer be needed, right?

**Comment by spalladino** [comment]
I've just stumbled upon [this bug](https://medium.com/@raiden_network/microraiden-bug-bounty-update-695fddd6433) from Microraiden. Are we subject to a similar issue?

> The way we were handling closing_sig was suboptimal. The closing_sig was a signature of the hash of balance_msg_sig. But that is not safe, as an ECDSA signature cannot replace a hash digest of a message. It does not provide non-repudiation if messages can be chosen before the signer’s private key. It is possible to construct a private key that will have the same signature for two specific messages.
> 
> After further investigation, we discovered that in our case, there was no possibility of using the exploit as presented in the above linked paper due to Ethereum’s ECDSA using v to differentiate between the possible curve points for the same signature.

More info in https://github.com/raiden-network/microraiden/issues/266

**Comment by AugustoL** [comment]
@spalladino ready for another review. 
Regarding the bug of microraiden I think It cannot happen here, but Im not sure, I will need to do more research.

File: contracts/token/ERC20/ERC20Channel.sol [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+  // The amount of time that a receiver has to challenge the sender
+  uint256 public challengeTime;
+
+  // The receiver of the tokens
+  address public receiver;
+
+  // The closing balance requested by the sender
+  uint256 public closingBalance;
+
+  // The timestamp of when the channel can ble closed by the sender
```
Comment (crypto-banana) [inline contracts/token/ERC20/ERC20Channel.sol]: `can be closed`

File: contracts/token/ERC20/ERC20Channel.sol [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+  uint256 public closeTime;
+
+  /**
+   * @dev Constructor
+   * @param tokenAddress address, the address of the ERC20 token that will be used
+   * @param _receiver address, the address of the receiver
+   * @param _challengeTime uint256, the time that a channel has before ends
+    with and uncoopertive close
```
Comment (crypto-banana) [inline contracts/token/ERC20/ERC20Channel.sol]: `with an uncooperative close`

File: contracts/token/ERC20/ERC20Channel.sol [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+   */
+
+  /**
+   * @dev Request an uncooperativeClose, it can be called only by the
+    sender/owner. It will save the closing balance requested and start the
+    challenge time period where the receiver can ask for a cooperativeClose.
+   * @param balance uint256, the final balance of the receiver
+   */
+  function uncooperativeClose(uint256 balance) external onlyOwner {
+    // Check that the closing request dont exist
```
Comment (crypto-banana) [inline contracts/token/ERC20/ERC20Channel.sol]: `request doesn't exist`

File: contracts/token/ERC20/ERC20Channel.sol [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    require(closeTime == 0);
+
+    // Check that the balance is less or equal of the token balance
```
Comment (crypto-banana) [inline contracts/token/ERC20/ERC20Channel.sol]: `equal to the token balance`

**Comment by AugustoL** [comment]
@crypto-banana thx for the review! typos fixed ^^

File: contracts/token/ERC20/ERC20Channel.sol [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+  /**
+   * @dev Close a channel
+   * @param finalBalance uint256, the final balance of the receiver
+   */
+  function close(uint256 finalBalance) internal {
+    uint256 tokenBalance = token.balanceOf(address(this));
+    require(finalBalance <= tokenBalance);
+
+    // Send balance to the receiver
+    require(token.transfer(receiver, finalBalance));
```
Comment (spalladino) [inline contracts/token/ERC20/ERC20Channel.sol]: Not sure if it's relevant, but if the token is an ERC223, the receiver could reenter at this point, presenting the same signed message from the sender, and extract twice the funds (or n times, if he reenters more than once). Or he could force the transfer to fail, making the closure of the channel impossible.

Would it be safer to change these `transfer` calls to `approvals`, so each user can later extract the balances themselves?

Comment (spalladino) [reply contracts/token/ERC20/ERC20Channel.sol]: On the other hand, given that both addresses need to sign messages with their private keys, they both are user addresses, and not contracts, so this sould not be a problem. Right?

Comment (spalladino) [reply contracts/token/ERC20/ERC20Channel.sol]: And also, I'm not sure if ERC223 is being much used at all

File: contracts/token/ERC20/ERC20Channel.sol [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+     Can happen at anytime while the channel is still opened. Two signatures
+     are needed for this, one is the signature of the receiver agreeing to
+     receive a certain value, with the receiver signature the sender can now
+     agree on the value and use sign the receiver signature.
+
+      cooperativeClose(
+       FinalBalance,
+       Receiver.sign(FinalBalance),
+       Sender.sign(SHA3( Receiver.sign(FinalBalance) ))
+      )
```
Comment (spalladino) [inline contracts/token/ERC20/ERC20Channel.sol]: I think that Receiver and Sender are swapped here, right?

File: test/token/ERC20/ERC20Channel.test.js [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    assert.equal(parseInt(channelInfo[1]), 0);
+    assert.equal(parseInt(channelInfo[2]), 0);
+    await tokenChannel.cooperativeClose(20, senderSig, closingSig, { from: receiver });
+    (await token.balanceOf(sender)).should.be.bignumber
+      .equal(80);
+    (await token.balanceOf(receiver)).should.be.bignumber
+      .equal(20);
+  });
+
+  it('create channel and close it with a mutual agreement from receiver', async function () {
```
Comment (spalladino) [inline test/token/ERC20/ERC20Channel.test.js]: WDTY about dynamically generating this test and the previous one using a loop over `[sender, receiver, anotherAddress]`, in order not to duplicate code?

File: test/token/ERC20/ERC20Channel.test.js [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  });
+
+  it('create channel and close it from sender with uncooperativeClose', async function () {
+    tokenChannel = await ERC20Channel.new(token.address, receiver, duration.days(1), { from: sender });
+    await token.transfer(tokenChannel.address, 30, { from: sender });
+    const hash = await tokenChannel.generateBalanceHash(10);
+    const senderSig = signMsg(hash, privateKeys[sender]);
+    assert.equal(sender,
+      await tokenChannel.getSignerOfBalanceHash(10, senderSig)
+    );
```
Comment (spalladino) [inline test/token/ERC20/ERC20Channel.test.js]: This signature is not needed, right?

File: test/token/ERC20/ERC20Channel.test.js [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    const closingSig = signMsg(senderHash, privateKeys[receiver]);
+    await tokenChannel.uncooperativeClose(10, { from: sender });
+    await increaseTimeTo(latestTime() + 10);
+    await tokenChannel.cooperativeClose(20, senderSig, closingSig, { from: receiver });
+    (await token.balanceOf(sender)).should.be.bignumber
+      .equal(80);
+    (await token.balanceOf(receiver)).should.be.bignumber
+      .equal(20);
+  });
+});
```
Comment (spalladino) [inline test/token/ERC20/ERC20Channel.test.js]: Please add tests/assertions for the following:
- Check that the contract is indeed destroyed after closing
- Trying to close the contract with a final balance greater than the collateral
- Sending incorrect/forged signatures in the cooperative close (signed with an incorrect private key, signed over a different balance, etc; for both the sender and the receiver signed messages)

File: contracts/token/ERC20/ERC20Channel.sol [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+   * @dev Constructor
+   * @param tokenAddress address, the address of the ERC20 token that will be used
+   * @param _receiver address, the address of the receiver
+   * @param _challengeTime uint256, the time that a channel has before ends
+    with an uncooperative close
+   */
+  function ERC20Channel(address tokenAddress, address _receiver, uint256 _challengeTime) public {
+    require(tokenAddress != address(0));
+    require(_receiver != address(0));
+    require(_challengeTime >= 0);
```
[the remark this answers is not in the document]
Comment (AugustoL) [reply contracts/token/ERC20/ERC20Channel.sol]: right! and also use a higher minimum challenge time maybe?? 10 minutes?

File: contracts/token/ERC20/ERC20Channel.sol [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+
+  /*
+   *  Public functions
+   */
+
+   /**
+    * @dev Get the channel info
+    */
+  function getInfo() public view returns (uint256, uint256, uint256) {
+    return (token.balanceOf(address(this)), closeTime, closingBalance);
```
[the remark this answers is not in the document]
Comment (AugustoL) [reply contracts/token/ERC20/ERC20Channel.sol]: The token balance gives you the maximun balance to be transfered.
The closeTime gives you the time for the channel to be closed by the sender if he requested it, if not it is 0.
The closingBalance gives you the final value requested by the sender if if he requested it, if not it is 0.
I think this values are the most important from the channel, thats why I used this ones, maybe I should add the sender and receiver address too?

File: contracts/token/ERC20/ERC20Channel.sol [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+   *  External functions
+   */
+
+  /**
+   * @dev Request an uncooperativeClose, it can be called only by the
+    sender/owner. It will save the closing balance requested and start the
+    challenge time period where the receiver can ask for a cooperativeClose.
+   * @param balance uint256, the final balance of the receiver
+   */
+  function uncooperativeClose(uint256 balance) external onlyOwner {
```
[the remark this answers is not in the document]
Comment (AugustoL) [reply contracts/token/ERC20/ERC20Channel.sol]: Good idea! that should happen!

File: contracts/token/ERC20/ERC20Channel.sol [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+
+  /**
+   * @dev Close a channel with the agreement of the sender and receiver
+   * @param balance uint256, the final balance transfered of the channel
+   * @param balanceMsgSig bytes, the signature of the sender.
+    The msg signed by the receiver is generateBalanceHash(balance)
+   * @param closingSig bytes, the signature of the receiver
+    The msg signed by the sender is generateKeccak256(balanceMsgSig)
+   */
+  function cooperativeClose(uint256 balance, bytes balanceMsgSig, bytes closingSig) external {
```
[the remark this answers is not in the document]
Comment (AugustoL) [reply contracts/token/ERC20/ERC20Channel.sol]: :+1:

File: contracts/token/ERC20/ERC20Channel.sol (hunk 13 above)
Comment (AugustoL) [reply contracts/token/ERC20/ERC20Channel.sol]: Yes, I agree too, is not being used much, but the problem still there, and it can be solved changing transfer with approval. Regarding user signatures that should be a problem, we should add a check that verifies that the sender and receiver are no contracts in the constructor. This way we will make sure it will be used by user addresses.
WDYT @spalladino ?

File: contracts/token/ERC20/ERC20Channel.sol (hunk 14 above)
Comment (AugustoL) [reply contracts/token/ERC20/ERC20Channel.sol]: I think it is fine, it is with the dame order n the `uncooperativeClose` function, the first signature is the one form the receiver and the second one from the sender.

File: test/token/ERC20/ERC20Channel.test.js (hunk 15 above)
Comment (AugustoL) [reply test/token/ERC20/ERC20Channel.test.js]: to select three random addresses from the 10 ones being used by testrpc? If it is that what you mean I think it will be better.

File: test/token/ERC20/ERC20Channel.test.js (hunk 16 above)
Comment (AugustoL) [reply test/token/ERC20/ERC20Channel.test.js]: it is not, killing it.

File: test/token/ERC20/ERC20Channel.test.js (hunk 17 above)
Comment (AugustoL) [reply test/token/ERC20/ERC20Channel.test.js]: :+1:

File: contracts/token/ERC20/ERC20Channel.sol [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+
+  /**
+   * @dev Get the signer of a balance hash signed
+   * @param balance uint256, the balance to hash
+   * @param msgSigned bytes, the balance hash signed
+   */
+  function getSignerOfBalanceHash(uint256 balance, bytes msgSigned) public view returns (address) {
+    bytes32 msgHash = generateBalanceHash(balance);
+    // Derive address from signature
+    address signer = msgHash.recover(msgSigned);
```
Comment (crypto-banana) [inline contracts/token/ERC20/ERC20Channel.sol]: Save a little gas and just return here instead of using a local variable?

File: test/token/ERC20/ERC20Channel.test.js (hunk 15 above)
Comment (spalladino) [reply test/token/ERC20/ERC20Channel.test.js]: Sorry for the confusion: I meant to dynamically generate the tests `create channel and close it with a mutual agreement from sender` and `create channel and close it with a mutual agreement from receiver` since they share most of the code. So, instead of having two tests that duplicate most of the code, do something like:

```js
[['sender', sender], ['receiver', receiver]].forEach(function(description, who) {
   it(`create channel and close it with a mutual agreement from ${description}`, async function () {
     tokenChannel = await ERC20Channel.new(token.address, receiver, duration.days(1), { from: sender });
     await token.transfer(tokenChannel.address, 30, { from: sender });
     const hash = await tokenChannel.generateBalanceHash(20);
     const senderSig = signMsg(hash, senderPrivateKey);
     assert.equal(sender,
       await tokenChannel.getSignerOfBalanceHash(20, senderSig)
     );
     const senderHash = await tokenChannel.generateKeccak256(senderSig);
     const closingSig = signMsg(senderHash, receiverPrivateKey);
     const channelInfo = await tokenChannel.getInfo();
     assert.equal(parseInt(channelInfo[0]), 30);
     assert.equal(parseInt(channelInfo[1]), 0);
     assert.equal(parseInt(channelInfo[2]), 0);
     await tokenChannel.cooperativeClose(20, senderSig, closingSig, { from: who });
     (await token.balanceOf(sender)).should.be.bignumber
       .equal(80);
     (await token.balanceOf(receiver)).should.be.bignumber
       .equal(20);
   });
});
```

Or define a helper function that runs the test:

```js
function createAndCloseChannelWithMutualAgreementFrom(description, who) {
  it(`create channel and close it with a mutual agreement from ${description}`, async function () {
    tokenChannel = await ERC20Channel.new(token.address, receiver, duration.days(1), { from: sender });
    await token.transfer(tokenChannel.address, 30, { from: sender });
    const hash = await tokenChannel.generateBalanceHash(20);
    const senderSig = signMsg(hash, senderPrivateKey);
    assert.equal(sender,
      await tokenChannel.getSignerOfBalanceHash(20, senderSig)
    );
    const senderHash = await tokenChannel.generateKeccak256(senderSig);
    const closingSig = signMsg(senderHash, receiverPrivateKey);
    const channelInfo = await tokenChannel.getInfo();
    assert.equal(parseInt(channelInfo[0]), 30);
    assert.equal(parseInt(channelInfo[1]), 0);
    assert.equal(parseInt(channelInfo[2]), 0);
    await tokenChannel.cooperativeClose(20, senderSig, closingSig, { from: who });
    (await token.balanceOf(sender)).should.be.bignumber
      .equal(80);
    (await token.balanceOf(receiver)).should.be.bignumber
      .equal(20);
  });
};

createAndCloseChannelWithMutualAgreementFrom('receiver', receiver);
createAndCloseChannelWithMutualAgreementFrom('sender', sender);
createAndCloseChannelWithMutualAgreementFrom('other', other);
```

That said, it's also interesting to have a random addresses helper. I'd add an option to seed it with a particular value though, so tests are reproducible.

File: contracts/token/ERC20/ERC20Channel.sol (hunk 13 above)
Comment (spalladino) [reply contracts/token/ERC20/ERC20Channel.sol]: Agree to change the transfer into an approval, which could even use an `asyncSend` version for ERC20s that we don't yet have. As for user signatures, I think it's ok not to check them, since the channel cannot be used if that were the case (no signatures can be exchanged). Besides, a user could spawn a contract in that address after the channel was created, making the check pointless.

File: contracts/token/ERC20/ERC20Channel.sol (hunk 19 above)
[the remark this answers is not in the document]
Comment (spalladino) [reply contracts/token/ERC20/ERC20Channel.sol]: Agree on adding those values!

File: contracts/token/ERC20/ERC20Channel.sol (hunk 13 above)
Comment (AugustoL) [reply contracts/token/ERC20/ERC20Channel.sol]: done!

**Comment by AugustoL** [comment]
@spalladino ready for another review, maybe the last one :)

File: contracts/token/ERC20/ERC20Channel.sol [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+   - Contract closed without agreement:
+    This can happen only in behalf of the sender, and it has a time to be
+     "challenged" by the receiver. The sender and receiver exchange signatures
+     of a final value to be transfered but they reach a point where they dont
+     agree on the final value. The sender can ask to close the channel with a
+     final value, if the challenge time passes and the channel does not receive
+     a cooperativeClose request it would be able to be closed by the sender
+     transfering the final value requested.
+ */
+contract ERC20Channel is Ownable {
```
Comment (spalladino) [inline contracts/token/ERC20/ERC20Channel.sol]: I've just realized that a sender can at any time `transferOwnership` of the contract, changing the sender and invalidating all previous signed messages,  since `getSignerOfBalanceHash` must be equal to the contract owner. So a malicious sender could transfer ownership of this contract to a different address he controls (after having sent several correctly signed payment messages with the original address), and run an uncooperative close to retrieve his funds. We'll need to remove the Ownable inheritance.

File: contracts/token/ERC20/ERC20Channel.sol (hunk 14 above)
Comment (spalladino) [reply contracts/token/ERC20/ERC20Channel.sol]: Hmm I still don't see it. The function is `cooperativeClose(uint256 balance, bytes balanceMsgSig, bytes closingSig)`, where closingSig is the *receiver's* signature of `balanceMsgSig`, which is the *sender's* signature of the hash of the balance. Am I missing sth?

File: contracts/token/ERC20/ERC20Channel.sol [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+
+  /**
+   * @dev Close a channel with the agreement of the sender and receiver
+   * @param balance uint256, the final balance transfered of the channel
+   * @param balanceMsgSig bytes, the signature of the sender.
+    The msg signed by the receiver is generateBalanceHash(balance)
+   * @param closingSig bytes, the signature of the receiver
+    The msg signed by the sender is generateKeccak256(balanceMsgSig)
+   */
+  function cooperativeClose(uint256 balance, bytes balanceMsgSig, bytes closingSig) external {
```
Comment (spalladino) [inline contracts/token/ERC20/ERC20Channel.sol]: I'm not sure if we already discussed this, but taking a look at an implementation for [ETH payment channels](https://github.com/machinomy/machinomy-contracts/blob/master/contracts/Unidirectional.sol), there is indeed a cooperative close function that can be invoked by the receiver directly, which removes the need for the closingSig in favour of checking that `msg.sender == receiver`. Should we add it here as well?

**Comment by nventuro** [comment]
Closing as stale. Also, we'd probably want to go for a more modular approach to state channels.
